library/pkparse.c

```diff
@@ -1153,10 +1282,29 @@ static int pk_parse_key_pkcs8_unencrypted_der(
 #endif /* MBEDTLS_RSA_C */
 #if defined(MBEDTLS_ECP_LIGHT)
     if (pk_alg == MBEDTLS_PK_ECKEY || pk_alg == MBEDTLS_PK_ECKEY_DH) {
-        if ((ret = pk_use_ecparams(&params, &mbedtls_pk_ec(*pk)->grp)) != 0 ||
-            (ret = pk_parse_key_sec1_der(mbedtls_pk_ec(*pk), p, len, f_rng, p_rng)) != 0) {
-            mbedtls_pk_free(pk);
-            return ret;
+        switch (ec_grp_id) {
+#if defined(MBEDTLS_ECP_DP_CURVE25519_ENABLED)
+            case MBEDTLS_ECP_DP_CURVE25519:
+#endif
+#if defined(MBEDTLS_ECP_DP_CURVE448_ENABLED)
+            case MBEDTLS_ECP_DP_CURVE448:
+#endif
+#if defined(MBEDTLS_ECP_DP_CURVE25519_ENABLED) || defined(MBEDTLS_ECP_DP_CURVE448_ENABLED)
+/* *INDENT-OFF* */
+                if ((ret = pk_use_ecparams_rfc8410(&params, ec_grp_id, &mbedtls_pk_ec(*pk)->grp)) != 0 ||
+                    (ret = pk_parse_key_rfc8410_der(mbedtls_pk_ec(*pk), p, len, end, f_rng, p_rng)) != 0) {
+                    mbedtls_pk_free(pk);
+                    return ret;
+                }
+                break;
+/* *INDENT-ON* */
+#endif
+            default:
+                if ((ret = pk_use_ecparams(&params, &mbedtls_pk_ec(*pk)->grp)) != 0 ||
+                    (ret = pk_parse_key_sec1_der(mbedtls_pk_ec(*pk), p, len, f_rng, p_rng)) != 0) {
+                    mbedtls_pk_free(pk);
+                    return ret;
+                }
```

Comment:
Pre-existing: we're not checking if there's anything else at the end of the buffer (after the `privateKey`). On one hand, there might be optional attributes or public key, which is valid, but there could be junk, making the whole  PrivateKeyInfo` syntactically invalid and we'd still accept it.

Reply:
If we can, we should probably have a test case with a key that has optional attributes and/or public key present.